### PR TITLE
remote: Flip clause for fast-forward only check

### PR DESCRIPTION
**Pull request by adityasaky**
In remote.updateLocalReferenceStorage, this commit updates a check to see if the reference is for a branch (in refs/heads) to checking the reference is not a tag. This change ensures that the subsequent fast-forward only handling clauses apply to references that are not standard branches stored in refs/heads.

I'm still missing tests etc here. If this fix makes sense, happy to work towards that (though I might need some pointers again).

#874

### CODE REVIEW

**Comment by adityasaky** [comment]
I tried a couple times to get the commit message right but looks like I'm still missing something. :(

**Comment by pjbgf** [comment]
> I tried a couple times to get the commit message right but looks like I'm still missing something. :(

That was actually my bad, I forgot to add `git` as a valid prefix. Do you mind [adding that](https://github.com/go-git/go-git/blob/master/.github/workflows/pr-validation.yml#L22) as a separate commit please?

File: remote.go [hunk 1]
```diff
@@ -1200,7 +1200,7 @@ func (r *Remote) updateLocalReferenceStorage(
 
 			// If the ref exists locally as a branch and force is not specified,
 			// only update if the new ref is an ancestor of the old
-			if old != nil && old.Name().IsBranch() && !force && !spec.IsForceUpdate() {
+			if old != nil && !old.Name().IsTag() && !force && !spec.IsForceUpdate() {
```
Comment (pjbgf) [inline remote.go:1203]: Please amend the comments above to align with the changes.

**Comment by pjbgf** [comment]
> I'm still missing tests etc here. If this fix makes sense, happy to work towards that (though I might need some pointers again).

As for tests, here's an example around fetching: https://github.com/go-git/go-git/blob/master/remote_test.go#L409.

**Comment by adityasaky** [comment]
@pjbgf I actually just added some tests for this but I couldn't work with the existing framework :cry:

But I think the case shows the change in behavior, please let me know if it makes sense.

**Comment by adityasaky** [comment]
> That was actually my bad, I forgot to add git as a valid prefix. Do you mind [adding that](https://github.com/go-git/go-git/blob/master/.github/workflows/pr-validation.yml#L22) as a separate commit please?

Done! That'll still fail because I didn't want to add my tests to my main commit until we figured out how to integrate them into the existing framework.

File: remote.go (hunk 1 above)
Comment (adityasaky) [reply remote.go:1203]: Done!

**Comment by adityasaky** [comment]
Re: https://github.com/go-git/go-git/issues/874#issuecomment-1762320375

I _think_ I hit this when I tried to use Clone with ReferenceName set to the custom ref in the test. At the time, I wasn't setting the master branch at all, but right now the test sets a master branch, clones the repo, then fetches the custom ref for the first time separately.

**Comment by adityasaky** [comment]
I squashed my tests commit to hopefully make CI go green. I still want to highlight that the test setup deviates a fair bit from the default test framework in the repository.

File: remote_test.go [hunk 2]
```diff
@@ -1555,3 +1559,140 @@ func (s *RemoteSuite) TestFetchAfterShallowClone(c *C) {
...
+		},
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// 4. Make divergent changes
+	remoteCommitID := writeCommitToRef(t, remoteRepo, customRef, emptyTreeID)
+	// Sleep so that consecutive calls to writeCommitToRef have different
+	// time.Now() values
+	time.Sleep(time.Second)
```
Comment (pjbgf) [inline remote_test.go:1600]: Instead of sleeping, we could send time as a parameter for `writeCommitToRef`.

```suggestion
```

Comment (adityasaky) [reply remote_test.go]: Done!

**Review by pjbgf — APPROVED** [review]
Thanks for working on this @adityasaky! :bow:

**Comment by adityasaky** [comment]
thank you!

**Pull request merged**